### PR TITLE
For joins that are too large to brute force reordering, use a hueristic reordering instead.

### PR DESCRIPTION
**Pull request by nicktobey**
This is kind of a hack, but is a sample of the kind of heuristics we could be doing to improve join planning.

I experimented with applying this "fast reordering" more generally, not just to really big queries. I suspect in the fast majority of cases it should give us what we want. But there were some existing planning tests that generated less optimal plans (replacing a lookup join with an unoptimized join.) This could potentially be improved, but my main priority is improving the performance of the slow test queries, so in the meantime this is limited to only running on those queries (joins on 20+ tables)

This passes a bunch of previously timing out sqllogictests and doesn't break any existing tests. I can add more tests before merging.

### CODE REVIEW

**Review by max-hoffman — COMMENTED** [review]
I like the fast path, using FDs to try to find lookup joins bottom-up is what MySQL and postgres does I think. The one thing that I'm confused about is the clarify around added plans. I was expecting you to try accumulating a full lookup plan selecting from a bag of edges, but if I'm reading correctly what this does is reorder groups where we know we'll find lookup joins specifically for the in-place order. The subset iteration might need a bitshift to do that, not sure if this currently works as expected. And if we know we have a lookup between two subgraphs, is reordering productive? Or is the addPlan logic just a sanity for making sure we aren't violating any left-join reordering?

**Comment by nicktobey** [comment]
Accumulating a full lookup plan selecting from a bag of edges is definitely doable, but would require a more complicated analysis, potentially aided by FD. Because we need to pick the edges in the right order. Selectively adding plans is much simpler for now.

> if I'm reading correctly what this does is reorder groups where we know we'll find lookup joins specifically for the in-place order. 

This isn't limited to the in-place order. We generate every possible plan that can be implemented as a trivial single-column lookup. We're still generating a number of plans that grows exponentially with the number of tables, but it's a smaller exponent.

I'm not expecting this to be a permanent solution as much as it is a possible direction to take reordering. The plans generated by this should probably always be preferable to other plans.

**Comment by max-hoffman** [comment]
Could you give some guidance on what groups we expect to add? For an edge between `S1` with one vertex, and some arbitrary `S2`, do we add `(S1, S2)`, and then every `(S1, S3)` where `S2` is a subset of `S3` (and `S3` doesn't intersect `S1`)? Is it possible to create orphaned groups that have no path to the root plan? Reiterating from above, do we need to reorder `(S1,S2)` or can we inline check and add `(S1, S2)` because we've already found the best lookup plan for that pair (other lookup options will be added independently by the algo)? The `join_order_builder_test.go` might be a good way to help me understand how this works better.

**Comment by nicktobey** [comment]
We expect to add every single `(S1, S2)` where either S1 or S2 has a single vertex.

Note that `addPlans` is the same method called by `dbSube`. By default, `dbSube` calls `addPlans` with every single possible pair of vertex sets, and it's the responsibility of `addPlans` to determine which plans are actually eligible, whether they can commute, etc. Calling `addPlans` doesn't guarentee that a join gets added, it's merely a prerequisite.

This approach has strictly fewer calls to `addPlans` and adds strictly fewer plans. Some of them might be orphaned, yes, but it's still less work.

**Comment by max-hoffman** [comment]
Would it be accurate to say that this is like "no bushy trees", and then filtered a second time for no new edges that haven't been verified as lookup joins?

File: sql/memo/join_order_builder.go [hunk 1]
```diff
@@ -459,6 +483,73 @@ func (j *joinOrderBuilder) dbSube() {
...
+// match an edge in the graph. This requires O(4^N) runtime on the number of relations and thus does not scale
+// for large joins. Thus, if the number of relations is above a threshold, we won't attempt to reorder.
+func (j *joinOrderBuilder) shouldUseFastReorder() bool {
+	return len(j.vertices) > 20
+}
+
+// fastReorder is an alternate implementation of dbSube that is optimized for joins on a large number of tables.
+// For these joins, a brute force approach is no longer practical. So instead, we only want to generate joins where
+// one side of the join is indexible. This will include the optimal plan for most joins with a small cardinality output.
+// (This particular implementation uses the FDs strict keys as an estimate for indexible columns, since that information
```
Comment (max-hoffman) [inline sql/memo/join_order_builder.go:499]: fds comment is maybe stale now

File: sql/memo/join_order_builder.go [hunk 2]
```diff
@@ -127,14 +127,16 @@ type joinOrderBuilder struct {
...
+	m                       *Memo
+	plans                   map[vertexSet]*ExprGroup
+	edges                   []edge
+	vertices                []RelExpr
+	vertexGroups            []GroupId
+	innerEdges              edgeSet
+	nonInnerEdges           edgeSet
+	newPlanCb               func(j *joinOrderBuilder, rel RelExpr)
+	indexibleCols           sql.ColSet
+	forceFastReorderForTest bool
```
Comment (max-hoffman) [inline sql/memo/join_order_builder.go:139]: maybe, `forceFastReorder`? `disableBushy`?

**Review by max-hoffman — APPROVED** [review]
Seems reasonable. Disable bushy plans for huge joins. If the join is well-formed we probably find a good lookup plan. Any customer that run into slow 20+ table joins can ask us if they need more.

**Comment by max-hoffman** [comment]
@nicktobey can we close?

**Comment by nicktobey** [comment]
Yeah this isn't really needed anymore because of https://github.com/dolthub/go-mysql-server/pull/2064, and didn't even have any asymptotic runtime improvement.

If we ever need to revisit this, we're better off replacing our current brute-force join planning with one that generates joins lazily, once we've determined they can actually show up in a final ordering, instead of the current implementation of visiting the power set of all tables and generating tons of plans that can't possible be in the final plan.

**Pull request closed without merging**